teps/0084-endtoend-provenance-collection.md

```diff
@@ -263,6 +505,34 @@ of the file, but general guidance is to include at least TEP number in the
 file name, for example, "/teps/images/NNNN-workflow.jpg".
 -->
 
+<!-- TODO: Revisit this once event-payload is taken into consideration -->
+
+### Pipelinerun Attestation
+
+The process of creating and storing pipelinerun attestations should mimic the process for taskrun
+attestations.
+
+In Tekton Chains, a new controller and a new reconciler are introduced to manage pipelinerun
+resources. Whenever a pipelinerun resource is updated, the reconciler is responsible for
+determining, and performing, any action to create and store a pipelinerun attestation.
+
+To avoid timing issues, it is important for the pipelinerun reconciler to only take action once all
+of the underlying taskrun resources have been reconciled. Otherwise, both the taskrun and
+pipelinerun reconcilers may create their corresponding attestation at the same time causing one of
+them to be lost.
```

Comment:
Would it be a valid consideration that if pipelinerun attestation is enabled, we disable the taskrun reconciler so that taskrun level attestation will not be generated at all? The main reason is that `predicate.buildConfig:` field already includes all the taskruns of the pipeline.
wdyt?

Reply:
If the TaskRun attestations are disabled, it also disables signing images based on TaskRuns. That's just how the implementation is done today.

In order to avoid unnecessarily signing the same image multiple times, I'm suggesting we do not sign images as part of the PipelineRun reconciler, and only use it to create attestations. So if you disable the TaskRun reconciler, you won't get image signatures. I think this is good for now.

In the future, I really think we should consider dropping the TaskRun reconciler altogether and just rely on PipelineRuns. At that point, we should consider moving the image signing to the PipelineRun reconciler. I do suggest that we postpone this change until PipelineRun attestations have been incorporated into the code base for a while.

Reply:
SGTM!